src/proof/supernova.rs

```diff
@@ -0,0 +1,598 @@
+    fn arity(&self) -> usize {
+        MultiFrame::<'_, F, C>::public_input_size() / 2
+    }
+
+    fn circuit_index(&self) -> usize {
+        self.multiframe.circuit_index()
+    }
+
+    fn synthesize<CS>(
+        &self,
+        cs: &mut CS,
+        pc: std::option::Option<&AllocatedNum<F>>,
+        z: &[AllocatedNum<F>],
+    ) -> Result<(std::option::Option<AllocatedNum<F>>, Vec<AllocatedNum<F>>), SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        if let Some(pc) = pc {
+            if pc.get_value() == Some(F::ZERO) {
+                debug!("synthesizing StepCircuit for Lurk");
+            } else {
+                debug!(
+                    "synthesizing StepCircuit for Coprocessor with pc: {:?}",
+                    pc.get_value()
+                );
+            }
+        }
+        let output = <MultiFrame<'_, F, C> as nova::traits::circuit::StepCircuit<F>>::synthesize(
+            &self.multiframe,
+            cs,
+            z,
+        )?;
+
+        let next_pc = AllocatedNum::alloc_infallible(&mut cs.namespace(|| "next_pc"), || {
+            self.next
+                .as_ref()
+                // This is missing in the case of a final `MultiFrame`. The Lurk circuit is defined to always have index
+                // 0, so it is a good default in this case.
+                .map_or(F::ZERO, |x| F::from(x.circuit_index() as u64))
+        });
+        debug!("synthesizing with next_pc: {:?}", next_pc.get_value());
+
+        Ok((Some(next_pc), output))
+    }
+}
+
+/// All steps of an NIVC computation
+pub struct NIVCSteps<'a, G: Group, C: Coprocessor<G::Scalar>>
+where
+    G::Scalar: LurkField,
+{
+    steps: Vec<NIVCStep<'a, G::Scalar, C>>,
+}
+impl<'a, G1, F, C1> Index<usize> for NIVCSteps<'a, G1, C1>
+where
+    C1: Coprocessor<F> + 'a,
+    G1: Group<Scalar = F> + 'a,
+    F: LurkField,
+{
+    type Output = NIVCStep<'a, G1::Scalar, C1>;
+
+    fn index(&self, idx: usize) -> &<Self as std::ops::Index<usize>>::Output {
+        &self.steps[idx]
+    }
+}
+
+impl<'a, G1, F, C1> NIVCSteps<'a, G1, C1>
+where
+    C1: Coprocessor<F>,
+    G1: Group<Scalar = F>,
+    F: LurkField,
+{
+    /// Number of NIVC steps contained.
+    pub fn num_steps(&self) -> usize {
+        self.steps.len()
+    }
+    /// Separate frames according to NIVC circuit requirements.
+    pub fn from_frames(
+        count: usize,
+        frames: &[Frame<IO<F>, Witness<F>, F, C1>],
+        store: &'a Store<F>,
+        folding_config: Arc<FoldingConfig<F, C1>>,
+    ) -> Self {
+        let mut steps = Vec::new();
+        let mut last_meta = frames[0].meta;
+        let mut consecutive_frames = Vec::new();
+
+        for frame in frames {
+            if frame.meta == last_meta {
+                let mut padding_frame = frame.clone();
+                // is this right, or should it be unmodified?
+                padding_frame.output = padding_frame.input;
```

Comment:
@porcuquine

Reply:
They are equivalent, since padding will only work if the frame is a no-op. This line can be removed (good catch) and was useful in development when unsuitable frames were still being emitted by reduction.

Reply:
#692 removes this.